keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/Stepsize.java

```diff
@@ -0,0 +1,165 @@
+    private static final double STARTING_STEPSIZE = 1;
+
+    private double stepsize;
+    private double averageTreeAcceptanceProb;
+    private double averageAcceptanceProb;
+    private double targetAcceptanceProb;
+    private double logStepSizeFrozen;
+    private double adaptCount;
+    private double shrinkageTarget;
+
+    /**
+     * @param stepsize                  the step size
+     * @param targetAcceptanceProb      the target acceptance probability (lower target equates to a higher step size when tuning)
+     * @param adaptCount                the number of samples to adapt for
+     */
+    Stepsize(double stepsize, double targetAcceptanceProb, int adaptCount) {
+        this.averageAcceptanceProb = 0.;
+        this.targetAcceptanceProb = targetAcceptanceProb;
+        this.stepsize = stepsize;
+        this.logStepSizeFrozen = Math.log(stepsize);
+        this.adaptCount = adaptCount;
+        this.shrinkageTarget = Math.log(10 * stepsize);
+        this.averageTreeAcceptanceProb = 0.;
+    }
+
+    /**
+     * Taken from algorithm 4 in https://arxiv.org/pdf/1111.4246.pdf.
+     *
+     * @param position                      the starting position
+     * @param gradient                      the gradient at the starting position
+     * @param vertices                      the vertices
+     * @param probabilisticVertices         the probabilistic vertices
+     * @param logProbGradientCalculator     the log prob gradient calculator
+     * @param initialLogOfMasterP           the initial master log prob
+     * @param random                        the source of randomness
+     *
+     * @return a starting step size
+     */
+    public static double findStartingStepSize(Map<VertexId, DoubleTensor> position,
+                                       Map<VertexId, DoubleTensor> gradient,
+                                       List<Vertex<DoubleTensor>> vertices,
+                                       List<Vertex> probabilisticVertices,
+                                       LogProbGradientCalculator logProbGradientCalculator,
+                                       double initialLogOfMasterP,
+                                       KeanuRandom random) {
+        double stepsize = STARTING_STEPSIZE;
+        Map<VertexId, DoubleTensor> momentums = new HashMap<>();
+        initializeMomentumForEachVertex(vertices, momentums, random);
+
+        Leapfrog leapfrog = new Leapfrog(position, momentums, gradient);
+        double pThetaR = initialLogOfMasterP - leapfrog.halfDotProductMomentum();
+
+        Leapfrog delta = leapfrog.step(vertices, logProbGradientCalculator, STARTING_STEPSIZE);
+
+        double probAfterLeapfrog = ProbabilityCalculator.calculateLogProbFor(probabilisticVertices);
+        double pThetaRAfterLeapFrog = probAfterLeapfrog - delta.halfDotProductMomentum();
+
+        double logLikelihoodRatio = pThetaRAfterLeapFrog - pThetaR;
+        double scalingFactor = logLikelihoodRatio > Math.log(0.5) ? 1 : -1;
```

Comment:
`signum` would be a better name, since it's always +1 or -1

Reply:
Agreed but we've kept the terminology the same as from the paper to make it easier to compare

Reply:
fair enough